bot/src/web/mod.rs

```diff
@@ -831,7 +834,13 @@ impl Api {
         };
 
         let c = player.current_device().await;
-        let data = player.list_devices().await?;
+        let data = match player.list_devices().await {
+            Ok(data) => data,
+            Err(_) => {
+                let data = Devices::default();
+                return Ok(warp::reply::json(&data));
+            }
+        };
```

Comment:
Excellent job so far. What about this one?

If there is an error code which indicates an empty collection of devices, it should be handled in the client instead. Otherwise this swallows the error.

Alternatively, this seems like it can just fall through to the newfangled error handling you added :)

Reply:
Spotify API sends an empty array when no devices are available.
The `Devices` component currently handles 0 devices as either missing authorization or actually no devices.
https://github.com/udoprog/OxidizeBot/blob/91afe3148c70ac5bd56d8b1a32d04cc581ca28fb/bot-ui/src/components/Devices.js#L87-L93
It would be possible to split these two situations, sending the error when authorization is missing.